{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e67575d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-6' coro=<TqApi._notify_watcher() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\api.py:3793> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-5' coro=<TqBaseApi._windows_patch() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py:146> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Exception ignored in: <coroutine object TqApi._notify_watcher at 0x000002956B1D1540>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\api.py\", line 3793, in _notify_watcher\n",
      "    async for _ in update_chan:\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\channel.py\", line 143, in __anext__\n",
      "    value = await asyncio.Queue.get(self)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\queues.py\", line 160, in get\n",
      "    getter.cancel()  # Just in case getter is not done yet.\n",
      "    ^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-7' coro=<TqConnect._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:181> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-8' coro=<TqReconnect._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:262> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-9' coro=<TqSymbols._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\symbols.py:29> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-10' coro=<TqTradingStatus._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\trading_status.py:26> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-11' coro=<TqConnect._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:181> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-12' coro=<TqReconnect._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:262> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-13' coro=<BaseOtg._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\tradeable\\otg\\base_otg.py:109> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-14' coro=<TqWebHelper._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\tqwebhelper.py:80> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-15' coro=<DataExtension._run() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\data_extension.py:85> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TqBaseApi._on_task_done()]>\n",
      "Exception ignored in: <coroutine object Connection.recv at 0x000002954E253340>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\connection.py\", line 294, in recv\n",
      "    return await self.recv_messages.get(decode)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\messages.py\", line 150, in get\n",
      "    frame = await self.frames.get(not self.closed)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\messages.py\", line 53, in get\n",
      "    self.get_waiter.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Exception ignored in: <coroutine object TqConnect._run at 0x000002954E02AF60>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\diff.py\", line 134, in _simple_merge_diff\n",
      "    for key in list(diff.keys()):\n",
      "               ^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: coroutine ignored GeneratorExit\n",
      "Exception ignored in: <coroutine object TqReconnect._run at 0x000002954E1A8EA0>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py\", line 312, in _run\n",
      "    await self._api._cancel_task(send_task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-16' coro=<TqReconnect._send_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:315> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object TqSymbols._run at 0x000002954E1A9010>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\symbols.py\", line 63, in _run\n",
      "    await self._api._cancel_task(sim_task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-17' coro=<TqSymbols._sim_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\symbols.py:67> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object TqTradingStatus._run at 0x000002954E1D9000>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\trading_status.py\", line 26, in _run\n",
      "    await super(TqTradingStatus, self)._run(api, api_send_chan, api_recv_chan, md_send_chan, md_recv_chan, self._ts_send_chan, self._ts_recv_chan)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py\", line 47, in _run\n",
      "    await self._api._cancel_tasks(*up_handle_tasks)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 58, in _cancel_tasks\n",
      "    await self._cancel_tasks(*other_tasks)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 55, in _cancel_tasks\n",
      "    await self._cancel_task(task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-19' coro=<TqModule._up_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py:50> wait_for=<Future cancelled>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-18' coro=<TqModule._up_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py:50> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object Connection.recv at 0x000002954E252240>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\connection.py\", line 294, in recv\n",
      "    return await self.recv_messages.get(decode)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\messages.py\", line 150, in get\n",
      "    frame = await self.frames.get(not self.closed)\n",
      "            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\messages.py\", line 53, in get\n",
      "    self.get_waiter.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Exception ignored in: <coroutine object TqConnect._run at 0x000002954E02B100>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\diff.py\", line 134, in _simple_merge_diff\n",
      "    for key in list(diff.keys()):\n",
      "               ^^^^^^^^^^^^^^^^^\n",
      "RuntimeError: coroutine ignored GeneratorExit\n",
      "Exception ignored in: <coroutine object TqReconnect._run at 0x000002954E1A9180>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py\", line 312, in _run\n",
      "    await self._api._cancel_task(send_task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-20' coro=<TqReconnect._send_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:315> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object BaseOtg._run at 0x000002954E1D9120>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\tradeable\\otg\\base_otg.py\", line 109, in _run\n",
      "    await super(BaseOtg, self)._run(api, api_send_chan, api_recv_chan, md_send_chan, md_recv_chan, self._td_send_chan, self._td_recv_chan)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py\", line 47, in _run\n",
      "    await self._api._cancel_tasks(*up_handle_tasks)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 58, in _cancel_tasks\n",
      "    await self._cancel_tasks(*other_tasks)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 55, in _cancel_tasks\n",
      "    await self._cancel_task(task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-22' coro=<TqModule._up_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py:50> wait_for=<Future cancelled>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-21' coro=<TqModule._up_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseModule.py:50> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object TqWebHelper._run at 0x00000295520A70B0>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\tqwebhelper.py\", line 85, in _run\n",
      "    await self._api._cancel_task(_data_handler_without_web_task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-23' coro=<TqWebHelper._data_handler_without_web() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\tqwebhelper.py:165> wait_for=<Future cancelled>>\n",
      "Exception ignored in: <coroutine object DataExtension._run at 0x000002954E250D40>\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\data_extension.py\", line 103, in _run\n",
      "    await self._api._cancel_task(md_task)\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 62, in _cancel_task\n",
      "    task.cancel()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\baseApi.py\", line 77, in _call_soon\n",
      "    return org_call_soon(callback, *args, **kargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 772, in call_soon\n",
      "    self._check_closed()\n",
      "  File \"c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\asyncio\\base_events.py\", line 519, in _check_closed\n",
      "    raise RuntimeError('Event loop is closed')\n",
      "RuntimeError: Event loop is closed\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-24' coro=<DataExtension._md_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\data_extension.py:107> wait_for=<Future cancelled>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-26' coro=<Connection.keepalive() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\connection.py:797> wait_for=<Future pending cb=[Task.task_wakeup()]>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-27' coro=<TqConnect._send_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:230> wait_for=<Future cancelled>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-28' coro=<Connection.keepalive() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\websockets\\asyncio\\connection.py:797> wait_for=<Future pending cb=[Task.task_wakeup()]>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task cancelling name='Task-29' coro=<TqConnect._send_handler() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\connect.py:230> wait_for=<Future cancelled>>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-34' coro=<TqDataFrame.async_update() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\objs_not_entity.py:195> wait_for=<Future finished result=None> cb=[TqBaseApi._on_task_done()]>\n",
      "Task was destroyed but it is pending!\n",
      "task: <Task pending name='Task-42' coro=<TqApi._get_serial_async() running at c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\api.py:755> cb=[TqBaseApi._on_task_done()]>\n",
      "c:\\Users\\96541\\anaconda3\\envs\\DeepLearning\\Lib\\site-packages\\tqsdk\\diff.py:134: RuntimeWarning: coroutine 'TqApi._get_serial_async' was never awaited\n",
      "  for key in list(diff.keys()):\n",
      "RuntimeWarning: Enable tracemalloc to get the object allocation traceback\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-167' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>\n",
      "RuntimeError: aclose(): asynchronous generator is already running\n",
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-168' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('aclose(): asynchronous generator is already running')>\n",
      "RuntimeError: aclose(): asynchronous generator is already running\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-27 23:11:28 -     INFO - 通知 x6504368: 与 wss://otg-sim.shinnytech.com/trade 的网络连接已建立\n",
      "2025-10-27 23:11:28 -     INFO - 通知 : 与 wss://free-api.shinnytech.com/t/nfmd/front/mobile 的网络连接已建立\n",
      "2025-10-27 23:11:28 -     INFO - 通知 x6504368: 登录成功\n"
     ]
    }
   ],
   "source": [
    "from tqsdk import TqApi, TqSim, TqAuth, TqKq\n",
    "import datetime\n",
    "import time\n",
    "tianqin = TqApi(TqKq(), auth=TqAuth('x6504368', 'q6504368'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "cf5482d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>datetime</th>\n",
       "      <th>id</th>\n",
       "      <th>open</th>\n",
       "      <th>high</th>\n",
       "      <th>low</th>\n",
       "      <th>close</th>\n",
       "      <th>volume</th>\n",
       "      <th>open_oi</th>\n",
       "      <th>close_oi</th>\n",
       "      <th>symbol</th>\n",
       "      <th>duration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2025-09-23 09:17:00</td>\n",
       "      <td>114659.0</td>\n",
       "      <td>855.20</td>\n",
       "      <td>855.40</td>\n",
       "      <td>855.02</td>\n",
       "      <td>855.02</td>\n",
       "      <td>1128.0</td>\n",
       "      <td>271078.0</td>\n",
       "      <td>271278.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2025-09-23 09:18:00</td>\n",
       "      <td>114660.0</td>\n",
       "      <td>855.02</td>\n",
       "      <td>855.04</td>\n",
       "      <td>854.66</td>\n",
       "      <td>854.98</td>\n",
       "      <td>1086.0</td>\n",
       "      <td>271278.0</td>\n",
       "      <td>271315.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2025-09-23 09:19:00</td>\n",
       "      <td>114661.0</td>\n",
       "      <td>854.98</td>\n",
       "      <td>855.04</td>\n",
       "      <td>854.70</td>\n",
       "      <td>854.88</td>\n",
       "      <td>838.0</td>\n",
       "      <td>271315.0</td>\n",
       "      <td>271408.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2025-09-23 09:20:00</td>\n",
       "      <td>114662.0</td>\n",
       "      <td>854.88</td>\n",
       "      <td>854.92</td>\n",
       "      <td>854.56</td>\n",
       "      <td>854.78</td>\n",
       "      <td>618.0</td>\n",
       "      <td>271408.0</td>\n",
       "      <td>271495.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2025-09-23 09:21:00</td>\n",
       "      <td>114663.0</td>\n",
       "      <td>854.78</td>\n",
       "      <td>854.80</td>\n",
       "      <td>854.12</td>\n",
       "      <td>854.12</td>\n",
       "      <td>864.0</td>\n",
       "      <td>271495.0</td>\n",
       "      <td>271473.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>2025-10-27 23:07:00</td>\n",
       "      <td>124654.0</td>\n",
       "      <td>913.60</td>\n",
       "      <td>913.64</td>\n",
       "      <td>912.66</td>\n",
       "      <td>913.16</td>\n",
       "      <td>1474.0</td>\n",
       "      <td>181155.0</td>\n",
       "      <td>181259.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>2025-10-27 23:08:00</td>\n",
       "      <td>124655.0</td>\n",
       "      <td>913.16</td>\n",
       "      <td>913.46</td>\n",
       "      <td>912.82</td>\n",
       "      <td>912.82</td>\n",
       "      <td>1223.0</td>\n",
       "      <td>181259.0</td>\n",
       "      <td>181434.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>2025-10-27 23:09:00</td>\n",
       "      <td>124656.0</td>\n",
       "      <td>912.82</td>\n",
       "      <td>912.92</td>\n",
       "      <td>912.30</td>\n",
       "      <td>912.76</td>\n",
       "      <td>1518.0</td>\n",
       "      <td>181434.0</td>\n",
       "      <td>181466.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>2025-10-27 23:10:00</td>\n",
       "      <td>124657.0</td>\n",
       "      <td>912.76</td>\n",
       "      <td>913.24</td>\n",
       "      <td>912.52</td>\n",
       "      <td>912.62</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>181466.0</td>\n",
       "      <td>181535.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>2025-10-27 23:11:00</td>\n",
       "      <td>124658.0</td>\n",
       "      <td>912.62</td>\n",
       "      <td>912.84</td>\n",
       "      <td>912.40</td>\n",
       "      <td>912.72</td>\n",
       "      <td>570.0</td>\n",
       "      <td>181535.0</td>\n",
       "      <td>181537.0</td>\n",
       "      <td>SHFE.au2512</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 datetime        id    open    high     low   close  volume  \\\n",
       "0     2025-09-23 09:17:00  114659.0  855.20  855.40  855.02  855.02  1128.0   \n",
       "1     2025-09-23 09:18:00  114660.0  855.02  855.04  854.66  854.98  1086.0   \n",
       "2     2025-09-23 09:19:00  114661.0  854.98  855.04  854.70  854.88   838.0   \n",
       "3     2025-09-23 09:20:00  114662.0  854.88  854.92  854.56  854.78   618.0   \n",
       "4     2025-09-23 09:21:00  114663.0  854.78  854.80  854.12  854.12   864.0   \n",
       "...                   ...       ...     ...     ...     ...     ...     ...   \n",
       "9995  2025-10-27 23:07:00  124654.0  913.60  913.64  912.66  913.16  1474.0   \n",
       "9996  2025-10-27 23:08:00  124655.0  913.16  913.46  912.82  912.82  1223.0   \n",
       "9997  2025-10-27 23:09:00  124656.0  912.82  912.92  912.30  912.76  1518.0   \n",
       "9998  2025-10-27 23:10:00  124657.0  912.76  913.24  912.52  912.62  1175.0   \n",
       "9999  2025-10-27 23:11:00  124658.0  912.62  912.84  912.40  912.72   570.0   \n",
       "\n",
       "       open_oi  close_oi       symbol  duration  \n",
       "0     271078.0  271278.0  SHFE.au2512        60  \n",
       "1     271278.0  271315.0  SHFE.au2512        60  \n",
       "2     271315.0  271408.0  SHFE.au2512        60  \n",
       "3     271408.0  271495.0  SHFE.au2512        60  \n",
       "4     271495.0  271473.0  SHFE.au2512        60  \n",
       "...        ...       ...          ...       ...  \n",
       "9995  181155.0  181259.0  SHFE.au2512        60  \n",
       "9996  181259.0  181434.0  SHFE.au2512        60  \n",
       "9997  181434.0  181466.0  SHFE.au2512        60  \n",
       "9998  181466.0  181535.0  SHFE.au2512        60  \n",
       "9999  181535.0  181537.0  SHFE.au2512        60  \n",
       "\n",
       "[10000 rows x 11 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def trans_time(ts):\n",
    "    try:\n",
    "        from datetime import datetime\n",
    "        timestamp_s = ts / 1e9\n",
    "        dt_object = datetime.fromtimestamp(timestamp_s)\n",
    "        formatted_time = dt_object.strftime(\"%Y-%m-%d %H:%M:%S\")\n",
    "        return formatted_time\n",
    "    except:\n",
    "        return ts\n",
    "data = tianqin.get_kline_serial('SHFE.au2512', duration_seconds=60, data_length=10000)\n",
    "df = data.copy().sort_values(\"datetime\")\n",
    "df['datetime'] = df['datetime'].apply(trans_time)\n",
    "# df = df[df['datetime']>=df[df['datetime'].str.contains(\" 21:00\")]['datetime'].iloc[0]]\n",
    "tianqin.wait_update()\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DeepLearning",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
